File: netlify/functions/upload-image.js
const { createClient } = require('@supabase/supabase-js');
const { GoogleGenerativeAI } = require('@google/generative-ai');

exports.handler = async (event, context) => {
  // Handle CORS preflight requests
  if (event.httpMethod === 'OPTIONS') {
    return {
      statusCode: 200,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type, Authorization',
        'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
      },
      body: ''
    };
  }
  
  // Only allow POST requests
  if (event.httpMethod !== 'POST') {
    return {
      statusCode: 405,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ error: 'Method not allowed' })
    };
  }
  
  try {
    // Parse request body
    const data = JSON.parse(event.body);
    
    if (!data.image || !data.user_id) {
      return {
        statusCode: 400,
        headers: {
          'Access-Control-Allow-Origin': '*',
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ error: 'Missing image or user_id' })
      };
    }
    
    const { image: imageData, user_id: userId, custom_prompt: customPrompt = 'Describe the contents of this image in detail. Be specific about objects, people, colors, and activities you see.' } = data;
    
    console.log('Function called with data:', { 
      hasImage: !!imageData, 
      imageLength: imageData?.length, 
      userId, 
      customPrompt: customPrompt?.substring(0, 50) + '...'
    });
    
    // Initialize Supabase client
    console.log('Creating Supabase client...');
    console.log('Environment variables available:', {
      SUPABASE_URL: !!process.env.SUPABASE_URL,
      SUPABASE_ANON_KEY: !!process.env.SUPABASE_ANON_KEY,
      GEMINI_API_KEY: !!process.env.GEMINI_API_KEY
    });
    
    const supabaseUrl = process.env.SUPABASE_URL || process.env.NEXT_PUBLIC_SUPABASE_URL;
    const supabaseKey = process.env.SUPABASE_ANON_KEY || process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
    
    if (!supabaseUrl || !supabaseKey) {
      console.error('Missing Supabase configuration:', { supabaseUrl: !!supabaseUrl, supabaseKey: !!supabaseKey });
      return {
        statusCode: 500,
        headers: {
          'Access-Control-Allow-Origin': '*',
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ 
          error: 'Supabase configuration missing',
          debug: {
            hasUrl: !!supabaseUrl,
            hasKey: !!supabaseKey
          }
        })
      };
    }
    
    const supabase = createClient(supabaseUrl, supabaseKey);
    
    // Initialize Gemini AI
    console.log('Creating Gemini AI client...');
    const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY);
    const model = genAI.getGenerativeModel({ model: 'gemini-1.5-flash' });
    
    // Store image in Supabase
    console.log('Storing image in database...');
    const { data: imageRecord, error: imageError } = await supabase
      .table('images')
      .insert({
        user_id: userId,
        image_data: imageData,
        status: 'processing'
      })
      .select()
      .single();
    
    if (imageError || !imageRecord) {
      console.error('Failed to store image:', imageError);
      return {
        statusCode: 500,
        headers: {
          'Access-Control-Allow-Origin': '*',
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ error: 'Failed to store image', details: imageError })
      };
    }
    
    console.log('Image stored with ID:', imageRecord.id);
    const imageId = imageRecord.id;
    
    // Convert base64 to proper format for Gemini
    console.log('Processing image for AI analysis...');
    const base64Data = imageData.split(',')[1];
    const mimeType = imageData.split(';')[0].split(':')[1];
    
    const imagePart = {
      inlineData: {
        data: base64Data,
        mimeType: mimeType,
      },
    };
    
    // Generate analysis using Gemini
    console.log('Calling Gemini AI...');
    const result = await model.generateContent([customPrompt, imagePart]);
    const analysisText = result.response.text();
    console.log('AI analysis completed, text length:', analysisText.length);
    
    // Store analysis results
    console.log('Storing analysis results...');
    const { data: analysisRecord, error: analysisError } = await supabase
      .table('analysis')
      .insert({
        image_id: imageId,
        user_id: userId,
        analysis_text: analysisText,
        custom_prompt: customPrompt,
        status: 'completed',
        is_edited: false
      })
      .select()
      .single();
    
    if (analysisError) {
      console.error('Failed to store analysis:', analysisError);
    }
    
    // Update image status
    console.log('Updating image status to completed...');
    await supabase
      .table('images')
      .update({ status: 'completed' })
      .eq('id', imageId);
    
    console.log('Function completed successfully');
    return {
      statusCode: 200,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        image_id: imageId,
        analysis: analysisText,
        analysis_id: analysisRecord?.id || null,
        status: 'completed'
      })
    };
    
  } catch (functionError) {
    console.error('Function error:', functionError);
    console.error('Error details:', {
      message: functionError.message,
      stack: functionError.stack,
      name: functionError.name
    });
    
    return {
      statusCode: 500,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ 
        error: `Function error: ${String(functionError)}`,
        details: {
          message: functionError.message,
          name: functionError.name
        }
      })
    };
  }
};

File: netlify/functions/upload-image.py
import json
import os
import base64
from PIL import Image
import io
import google.generativeai as genai
from supabase import create_client, Client

def handler(event, context):
    # Handle CORS preflight requests
    if event['httpMethod'] == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': {
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Headers': 'Content-Type, Authorization',
                'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
            },
            'body': ''
        }
    
    # Only allow POST requests
    if event['httpMethod'] != 'POST':
        return {
            'statusCode': 405,
            'headers': {
                'Access-Control-Allow-Origin': '*',
                'Content-Type': 'application/json',
            },
            'body': json.dumps({'error': 'Method not allowed'})
        }
    
    try:
        # Parse request body
        data = json.loads(event['body'])
        
        if 'image' not in data or 'user_id' not in data:
            return {
                'statusCode': 400,
                'headers': {
                    'Access-Control-Allow-Origin': '*',
                    'Content-Type': 'application/json',
                },
                'body': json.dumps({'error': 'Missing image or user_id'})
            }
        
        image_data = data['image']
        user_id = data['user_id']
        custom_prompt = data.get('custom_prompt', 'Describe the contents of this image in detail. Be specific about objects, people, colors, and activities you see.')
        
        # Initialize Supabase client
        supabase_url = os.getenv('SUPABASE_URL')
        supabase_key = os.getenv('SUPABASE_ANON_KEY')
        
        if supabase_url and supabase_key:
            supabase = create_client(supabase_url, supabase_key)
            
            # Store image in Supabase
            image_record = supabase.table('images').insert({
                'user_id': user_id,
                'image_data': image_data,
                'status': 'processing'
            }).execute()
            
            if not image_record.data:
                return {
                    'statusCode': 500,
                    'headers': {
                        'Access-Control-Allow-Origin': '*',
                        'Content-Type': 'application/json',
                    },
                    'body': json.dumps({'error': 'Failed to store image'})
                }
            
            image_id = image_record.data[0]['id']
        else:
            # Mock image ID for testing without Supabase
            image_id = 1
            supabase = None
        
        # Initialize Gemini model
        gemini_api_key = os.getenv('GEMINI_API_KEY')
        if not gemini_api_key:
            return {
                'statusCode': 500,
                'headers': {
                    'Access-Control-Allow-Origin': '*',
                    'Content-Type': 'application/json',
                },
                'body': json.dumps({'error': 'GEMINI_API_KEY not configured'})
            }
        
        genai.configure(api_key=gemini_api_key)
        model = genai.GenerativeModel('gemini-1.5-flash')
        
        # Process image with Gemini
        try:
            # Decode base64 image
            image_bytes = base64.b64decode(image_data.split(',')[1])
            
            # Convert to PIL Image
            img = Image.open(io.BytesIO(image_bytes))
            
            # Generate analysis using Gemini with custom prompt
            response = model.generate_content([custom_prompt, img])
            analysis_text = response.text
            
            # Store analysis results (if Supabase is configured)
            if supabase:
                analysis_record = supabase.table('analysis').insert({
                    'image_id': image_id,
                    'user_id': user_id,
                    'analysis_text': analysis_text,
                    'custom_prompt': custom_prompt,
                    'status': 'completed',
                    'is_edited': False
                }).execute()
                
                # Update image status
                supabase.table('images').update({
                    'status': 'completed'
                }).eq('id', image_id).execute()
            
            return {
                'statusCode': 200,
                'headers': {
                    'Access-Control-Allow-Origin': '*',
                    'Content-Type': 'application/json',
                },
                'body': json.dumps({
                    'image_id': image_id,
                    'analysis': analysis_text,
                    'analysis_id': analysis_record.data[0]['id'] if supabase and analysis_record.data else None,
                    'status': 'completed'
                })
            }
            
        except Exception as ai_error:
            # Update image status to failed (if Supabase is configured)
            if supabase:
                supabase.table('images').update({
                    'status': 'failed',
                    'error_message': str(ai_error)
                }).eq('id', image_id).execute()
            
            return {
                'statusCode': 500,
                'headers': {
                    'Access-Control-Allow-Origin': '*',
                    'Content-Type': 'application/json',
                },
                'body': json.dumps({'error': f'AI analysis failed: {str(ai_error)}'})
            }
            
    except Exception as e:
        return {
            'statusCode': 500,
            'headers': {
                'Access-Control-Allow-Origin': '*',
                'Content-Type': 'application/json',
            },
            'body': json.dumps({'error': str(e)})
        }